torch/_higher_order_ops/scan.py

```diff
@@ -4,32 +4,93 @@
 from typing import Any, Callable, List, Tuple
 
 import torch
+import torch._dynamo.variables
 import torch._prims_common as utils
 import torch._subclasses.functional_tensor
 import torch.utils._pytree as pytree
 from torch._C import DispatchKey
+from torch._dispatch.python import suspend_functionalization
 from torch._higher_order_ops.utils import (
     _has_potential_branch_input_alias,
     _has_potential_branch_input_mutation,
     _set_compilation_env,
-    autograd_not_implemented,
     reenter_make_fx,
     unique_graph_id,
     UnsupportedAliasMutationException,
 )
 from torch._ops import HigherOrderOperator
 from torch._subclasses.fake_tensor import FakeTensorMode
+from torch._subclasses.functional_tensor import disable_functional_mode
 from torch.fx.experimental.proxy_tensor import (
     disable_proxy_modes_tracing,
     ProxyTorchDispatchMode,
     track_tensor_tree,
 )
 from torch.utils._python_dispatch import _get_current_dispatch_mode
 
+from .utils import _from_fun, _maybe_reenter_make_fx, create_fw_bw_graph
+
 
 aten = torch._ops.ops.aten
 
 
+# Helper functions that are also used from other places
+def first_slice_copy(t: torch.Tensor, dim: int) -> torch.Tensor:
+    return torch.select_copy(t, dim, 0)
+
+
+def _extract_carry_and_out(flat_out: List[Any], num_carry: int):
+    return list(flat_out[:num_carry]), list(flat_out[num_carry:])
+
+
+# We also do a clone with contiguous_format. This is to be consistent with
+# eager semantic of scan, which stacks the outputs. The result is contiguous
+# as a result of the stack operation.
+def stack_y(y: torch.Tensor, scan_length: int) -> torch.Tensor:
+    return (
+        y.unsqueeze(0)
+        .repeat(*([scan_length] + [1] * y.ndim))
+        .clone(memory_format=torch.contiguous_format)
+    )
+
+
+# say we have a tensor of shape [3, 4, 5, 6]
+# shift_source_dim_to_target_dim(t, 0, 3) -> [4, 5, 6, 3]
+def shift_source_dim_to_target_dim(t, from_dim: int, to_dim: int):
+    assert to_dim >= 0 and to_dim < t.ndim
+    assert from_dim >= 0 and from_dim < t.ndim
+    dims = list(range(0, t.ndim))
+    dims.pop(from_dim)
+    dims.insert(to_dim, from_dim)
+    return t.permute(*dims)
+
+
+def prepare_xs_carries_for_bwd(xs, init, carries, dim, reverse):
```

Comment:
We don't expect two prepare_* functions to be used by other file, right? So maybe move these two prepare function inside of backward to make it locally scoped? They are not exposed to users in this way.

Reply:
I moved the functions inside the backward